executor/merge_join.go

```diff
@@ -172,36 +205,49 @@ func (t *mergeJoinInnerTable) hasNullInJoinKey(row chunk.Row) bool {
 	return false
 }
 
-// reallocReaderResult resets "t.curResult" to an empty Chunk to buffer the result of "t.reader".
-// It pops a Chunk from "t.resourceQueue" and push it into "t.resultQueue" immediately.
-func (t *mergeJoinInnerTable) reallocReaderResult() {
-	if !t.curResultInUse {
+// reallocCurChkForNext resets "t.curChk" to an empty Chunk to buffer the result of "t.reader".
+// It saves the curChk to RowContainer and then allocates a new one from RowContainer.
+func (t *mergeJoinInnerTable) reallocCurChkForReader() (err error) {
+	if !t.curChkInUse || t.curSel == nil {
 		// If "t.curResult" is not in use, we can just reuse it.
-		t.curResult.Reset()
+		// Note: t.curSel should never be nil. There is a case the chunk is in use but curSel is nil,
+		// and it would cause panic, so we add the condition here to avoid it. The case is that when
+		// init is called, and the firstRow4Key is set up by nextRow(), however the first several rows
+		// contain null value and are skipped, and in the next time the reallocCurChkFOrReader is called
+		// the curChkInUse would be set however the curSel is still nil.
+		t.curChk.Reset()
 		return
 	}
 
-	// Create a new Chunk and append it to "resourceQueue" if there is no more
-	// available chunk in "resourceQueue".
-	if len(t.resourceQueue) == 0 {
-		newChunk := newFirstChunk(t.reader)
-		t.memTracker.Consume(newChunk.MemoryUsage())
-		t.resourceQueue = append(t.resourceQueue, newChunk)
-	}
+	// Allocate a new Chunk from RowContainer
+	newChk := t.sameKeyRows.AllocChunk()
```

Comment:
Why not just use `newFirstChunk`? Since `newChk` is not obtained by `sameKeyRows`.

Reply:
Because we wanna reuse some chunks in the rowContainer. If the last time the RC has allocated some chunks, and it did not spill, then Reset the RC won't GC all these chunks, and next time AllocChunk will reuse these chunks.